### DISCUSSION TITLE
Implement constraints

### DISCUSSION THREAD

**Discussion by jdittrich** [uxbox/uxbox#533] [Ideas]
**Story:** 
* As a designer, I want to iterate quickly and change the dimensions of the elements on screen easily and without them becoming useless due to the resizing.
* In a team with devs and designers working together, it is helpful if their tools can emulate similar behaviors (Here: UI elements have a scaling behavior in UI toolkits and CSS/HTML which should be reflected in design tools)  

**Problem:** Currently, artboards, groups and their children are scaled uniformly. But e.g. a circle should stay always a circle (not become an ellipse or not scale at all) and a container for text should scale along instead. 

![example of a uniformly scaled image and a nicely scaled one (with constraints)](https://user-images.githubusercontent.com/3416487/106651399-62acf780-6594-11eb-86ab-c619c54e0506.png)

**Solution:** [Figma](https://www.figma.com/best-practices/tips-for-using-constraints-in-your-workflow/) and [Affinity Designer](https://affinity.help/designer/en-US.lproj/pages/DesignAids/constraints.html) implement "constraints". They allow to define for an element how it should behave if the parent element’s dimensions are changed. The most simple version is defining for each of the four edges of a rectangle if they should scale proportionally or always keep the same distance to the parent's edges.

![screenshot figma](https://user-images.githubusercontent.com/3416487/106958949-54004500-673a-11eb-9bcd-5702c61d54b3.png)
Figmas solution is compact and can be controlled completely via the two dropdowns.

![grafik](https://user-images.githubusercontent.com/3416487/106959285-bfe2ad80-673a-11eb-8f7b-704696bda63a.png)
Affinity designer has some more options, as you can influence scaling and relation to the edges independently and additionally to prefer enlarging or shrinking (effectively setting the current size as min- or max- dimension). It takes more space and more time to understand, though.

**To consider** 
* Needs some space in the UI (probably right hand sidebar)
* Not useful if users use penpot merely to create a "finished" design; constraints are most useful in smaller design iterations and/or when used with a design system and components.

**Comment by elhombretecla** [comment]
Thank you very much for your complete and meaningful feedback. We have prepared a rework of the components and groups with advanced  features like this. However, we will complete our User Story with your proposal. It is very useful!

**Comment by myfunnyandy** [comment]
Released! https://help.penpot.app/user-guide/constraints/
